File: Data Structures & Algorithms/trapping-rain-water/submission-4.py
class Solution:
    def trap(self, height: List[int]) -> int:
        prefix = [x for x in height]
        suffix = [x for x in height]
        n = len(height)

        for i in range(1,n):
            prefix[i]=max(prefix[i-1],prefix[i])
        for i in range(n-2,-1,-1):
            suffix[i]=max(suffix[i+1],suffix[i])
        ans = 0
        for i in range(1,n-1):
            ans += max(0,min(prefix[i-1],suffix[i+1])-height[i])
        return ans

File: Data Structures & Algorithms/buy-and-sell-crypto/submission-4.js
class Solution {
    /**
     * @param {number[]} prices
     * @return {number}
     */
    maxProfit(prices) {
        let min = prices[0]
        let ans = 0
        for (let i=0; i<prices.length; i++) {
            min = Math.min(min,prices[i])
            ans = Math.max(prices[i] - min, ans)
        }
        return ans
    }
}
